File: pascal/VMTranslator/VMTranslator.pas
program VMTranslator;

uses
  SysUtils,
  Parser, CodeWriter;

var
  Argument: string;
  InputFileName: string;
  OutputFileName: string;
  InputFile: Text;
  OutputFile: Text;
  Command: TCommand;
  arg1: string;
  arg2: string;
  Temp, Name, FunctionName, ModuleName: string;
  WorkingDir: string;
  FindPath: string;
  Info: TSearchRec;
  n: integer;
  IsDir: boolean;
begin
  if ParamCount = 1 then
  begin
    Argument := ParamStr(1);
    IsDir := DirectoryExists(Argument);
    if IsDir then
    begin
      while Argument[length(Argument)] = '/' do
        Argument := copy(Argument, 1, length(Argument) - 1);
      FindPath := '*.vm';
      Name := ExtractFileName(Argument);
      OutputFileName := ExpandFileName(Name + '.asm');
      WorkingDir := Argument;
    end
    else
    begin
      FindPath := Argument;
      Temp := ExtractFileName(Argument);
      Name := copy(Temp, 1, LastDelimiter('.', Temp) - 1);
      OutputFileName := ExpandFileName(ChangeFileExt(Name, '.asm'));
      WorkingDir := ExtractFileDir(Argument);
    end;
    
    writeln('Name=', Name);
    SetCurrentDir(WorkingDir);
    writeln('WorkingDir=', WorkingDir);
    writeln('OutputFileName=', OutputFileName);

    if FindFirst(FindPath, faAnyFile, Info) = 0 then
    begin
      InitWriter(OutputFileName, OutputFile);
      if IsDir then
        WriteInit(OutputFile, 0);
      repeat
        InputFileName := Info.Name;
        writeln('InputFileName=', InputFileName);

        ModuleName := copy(InputFileName, 1,
                           LastDelimiter('.', InputFileName) - 1);

        n := 1;
	FunctionName := 'Sys.init';
        InitParser(InputFileName, InputFile);
        while Advance(InputFIle, Command, arg1, arg2) do
        begin
          case Command of
            cArithm: WriteArithm(OutputFile, n, ModuleName, arg1);
            cPush: WritePush(OutputFile, ModuleName, arg1, arg2);
            cPop: WritePop(OutputFile, ModuleName, arg1, arg2);
            cGoto: WriteGoto(OutputFile, FunctionName, arg1);
            cLabel: WriteLabel(OutputFile, FunctionName, arg1);
            cIfGoto: WriteIfGoto(OutputFile, FunctionName, arg1);
            cReturn: WriteReturn(OutputFile, Name);
            cCall: WriteCall(OutputFile, n, arg1, StrToInt(arg2));
            cFunction: 
            begin
              FunctionName := arg1;
              WriteFunction(OutputFile, arg1, StrToInt(arg2))
            end
          end;
          n := n + 1
        end;
        CloseParser(InputFile);
      until FindNext(Info) <> 0;
      FindClose(Info);
      WriteEnd(OutputFile, Name);
      WriteRestore(OutputFile, Name);
      CloseWriter(OutputFile)
    end
  end
end.

File: java/VMTranslator/VMTranslator.java
import java.io.*;

public class VMTranslator {

    public static String getNoExtName(File file) {
        String name = file.getName();
        return name.substring(0, name.lastIndexOf('.'));
    }

    public static void main(String[] args) throws IOException {
        boolean writeEnv;
        if (args.length != 1) {
            System.out.println("usage: java VMTranslator <file/dir>");
            return;
        }

        /* get file and dir */
        File file = new File(args[0]);
        if (!file.exists()) {
            System.out.println(args[0] + " does not exist");
        }
        File dir;

        /* get file list and program name*/
        File[] inFiles;
        FilenameFilter vmFilter = new FilenameFilter() {
            public boolean accept(File dir, String name) {
                return name.endsWith(".vm");
            }
        };
        String programName;
        if (file.isDirectory()) {
            programName = file.getName();
            inFiles = file.listFiles(vmFilter);
            dir = file.getAbsoluteFile();
        } else {
            inFiles = new File[1];
            inFiles[0] = file;
            programName = getNoExtName(file);
            dir = file.getAbsoluteFile().getParentFile();
        }
        writeEnv = new File(dir, "Sys.vm").exists();

        File outFile = new File(dir, programName + ".asm");
//        System.out.println("output file = " + outFile.getCanonicalPath());
        CodeWriter codeWriter = new CodeWriter(outFile);
        if (writeEnv) {
            codeWriter.writeInit();
        }
        for (File f : inFiles) {
            String moduleName = getNoExtName(f);
            codeWriter.setModuleName(moduleName);
            Parser parser = new Parser(f);
            String[] strs;
            while ((strs = parser.advance()) != null) {
                doCmd(codeWriter, strs);
            }
            parser.close();
        }
        codeWriter.writeEndLoop();
        codeWriter.writeRestore();
        codeWriter.close();
    }

    private static void doCmd(CodeWriter cw, String[] strs) {
        switch (strs[0]) {
        case "push":
            cw.writePush(strs[1], Integer.parseInt(strs[2]));
            break;
        case "pop":
            cw.writePop(strs[1], Integer.parseInt(strs[2]));
            break;
        case "label":
            cw.writeLabel(strs[1]);
            break;
        case "goto":
            cw.writeGoto(strs[1]);
            break;
        case "if-goto":
            cw.writeIfGoto(strs[1]);
            break;
        case "return":
            cw.writeReturn();
            break;
        case "call":
            cw.writeCall(strs[1], Integer.parseInt(strs[2]));
            break;
        case "function":
            cw.writeFunction(strs[1], Integer.parseInt(strs[2]));
            break;
        default:
            cw.writeArithm(strs[0]);
            break;
        }
        cw.incN();
    }

}
